kratos/utilities/assign_mpcs_to_neighbours_utility.h

```diff
@@ -0,0 +1,252 @@
+          double const Radius,
+          ResultNodesContainerType& rResults);
+
+      /**
+       * @brief Get the component variable for a given vector variable and index
+       * @details Returns the component variable (e.g., DISPLACEMENT_X) corresponding to the input vector variable (e.g., DISPLACEMENT) and component index (0, 1, or 2).
+       * @param rVectorVariable Input vector variable of type array_1d<double, 3>.
+       * @param ComponentIndex Index of the component in the vector variable (0, 1, or 2).
+       * @return The component variable corresponding to the given vector variable and index.
+       **/
+      const Variable<double>& GetComponentVariable(
+          const Variable<array_1d<double, 3>>& rVectorVariable, 
+          const std::size_t ComponentIndex);
+
+
+      /**
+       * @brief Collect Dofs and Coordinates
+       * @details Collects Dofs and Coordinates for a given node or set of nodes.
+       * @param rStructureNodes Nodes Container.
+       * @param nodes_array Nodes Container.
+       * @param pNode Obtain respective dofs and coordinates for a given node or set of nodes.
+       * @param rVariable Dof variable array or double.
+       * @param rCloud_of_dofs Dofs container.
+       * @param rCloud_of_nodes_coordinates Coordinates container.
+       **/
+      void GetDofsAndCoordinatesForNode(
+        NodeType::Pointer pNode,
+        const Variable<double>& rVariable,
+        DofPointerVectorType& rCloud_of_dofs,
+        array_1d<double,3>& rSlave_coordinates
+        );
+
+      void GetDofsAndCoordinatesForNode(
+        NodeType::Pointer pNode,
+        const Variable<array_1d<double, 3>>& rVariable,
+        DofPointerVectorType& rCloud_of_dofs_x,
+        DofPointerVectorType& rCloud_of_dofs_y,
+        DofPointerVectorType& rCloud_of_dofs_z,
+        array_1d<double,3>& rSlave_coordinates
+        );
+
+      // Get Dofs and Coordinates arrays for a given variable double. (For nodes)
+      void GetDofsAndCoordinatesForNodes(
+        ResultNodesContainerType nodes_array,
+        const Variable<double>& rVariable,
+        DofPointerVectorType& rCloud_of_dofs,
+        Matrix& rCloud_of_nodes_coordinates
+        );
+
+      void GetDofsAndCoordinatesForNodes(
+        ResultNodesContainerType nodes_array,
+        const Variable<array_1d<double, 3>>& rVariable,
+        DofPointerVectorType& rCloud_of_dofs_x,
+        DofPointerVectorType& rCloud_of_dofs_y,
+        DofPointerVectorType& rCloud_of_dofs_z,
+        Matrix& rCloud_of_nodes_coordinates
+        );
+
+      /**
+       * @brief Assign rotation to nodes
+       * @details Assigns the given rotation matrix to the nodes' positions and updates their coordinates.
+       * @param pNodes Nodes container with the nodes to apply the rotation.
+       * @param RotationMatrix 3x3 rotation matrix to be applied to the nodes' positions.
+       **/
+      void AssignRotationToNodes(
+          NodesContainerType pNodes,
```

Comment:
This should be a reference. Indeed it is strange to me that this works if you pass it by copy...

Reply:
Not used anymore.